casr/src/triage.rs

```diff
@@ -0,0 +1,464 @@
+    if output_dir.read_dir()?.count() < 2 {
+        info!("There are less than 2 CASR reports, nothing to deduplicate.");
+        return summarize_results(matches, crashes);
+    }
+    info!("Deduplicating CASR reports...");
+    let casr_cluster_d = Command::new(&casr_cluster)
+        .arg("-d")
+        .arg(output_dir.clone().into_os_string())
+        .output()
+        .with_context(|| format!("Couldn't launch {casr_cluster:?}"))?;
+
+    if casr_cluster_d.status.success() {
+        info!(
+            "{}",
+            String::from_utf8_lossy(&casr_cluster_d.stdout)
+                .lines()
+                .collect::<Vec<&str>>()
+                .join(". ")
+        );
+    } else {
+        bail!("{}", String::from_utf8_lossy(&casr_cluster_d.stderr));
+    }
+
+    if !matches.get_flag("no-cluster") {
+        if output_dir
+            .read_dir()?
+            .flatten()
+            .map(|e| e.path())
+            .filter(|e| e.extension().is_some() && e.extension().unwrap() == "casrep")
+            .count()
+            < 2
+        {
+            info!("There are less than 2 CASR reports, nothing to cluster.");
+            return summarize_results(matches, crashes);
+        }
+        info!("Clustering CASR reports...");
+        let casr_cluster_c = Command::new(&casr_cluster)
+            .arg("-c")
+            .arg(output_dir.clone().into_os_string())
+            .output()
+            .with_context(|| format!("Couldn't launch {casr_cluster:?}"))?;
+
+        if casr_cluster_c.status.success() {
+            info!(
+                "{}",
+                String::from_utf8_lossy(&casr_cluster_c.stdout).trim_end()
+            );
+        } else {
+            error!("{}", String::from_utf8_lossy(&casr_cluster_c.stderr));
+        }
+
+        // Remove reports from deduplication phase. They are in clusters now.
+        for casrep in fs::read_dir(output_dir)?.flatten().map(|e| e.path()) {
+            if let Some(ext) = casrep.extension() {
+                if ext == "casrep" {
+                    let _ = fs::remove_file(casrep);
+                }
+            }
+        }
+    }
+
+    summarize_results(matches, crashes)
+}
+
+/// Copy crashes next to reports and print summary.
+/// Run casr-gdb on uninstrumented binary if specified in ARGS.
+/// Print analysis statistic.
+///
+/// # Arguments
+///
+/// * `matches` - tool arguments
+///
+/// * `crashes` - set of crashes, specified as a CrashInfo structure
+fn summarize_results(
+    matches: &clap::ArgMatches,
+    crashes: &HashMap<String, CrashInfo>,
+) -> Result<()> {
+    // Get output dir
+    let dir = matches.get_one::<PathBuf>("output").unwrap();
+    // Copy crashes next to reports
+    copy_crashes(dir, crashes)?;
+
+    // Get casr-gdb args
+    let tool = std::env::current_exe()?
+        .file_name()
+        .unwrap()
+        .to_str()
+        .unwrap()
+        .to_string();
+    let gdb_args = if tool.eq("casr-libfuzzer") {
```

Comment:
Maybe, we'd better pass gdb_args as a function argument and parse it in corresponding binary. Consider that there can appear casr-honggfuzz or similar

Reply:
Is it neccessary? We have all information in matches. In the case of honggfuzz, we can just add one check (`tool.eq("casr-honggfuzz")`) and everything will work. I removed gdb_args from function arguments a few commits ago.

Reply:
I suppose function should not check who called it. It's better to pass data straightforward.

Reply:
However you may leave the code in this function

Reply:
I have rewritten the code so we now parse gdb_args in the corresponding tools and pass them as a function argument.